Fixes #24280: When we clone a technique with resource, the clone does not really have the resource

VinceMacBuche opened the pull request:
https://issues.rudder.io/issues/24280

amousset commented on webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/TechniqueApi.scala:
```diff
@@ -213,6 +214,28 @@ class TechniqueApi(
+                 authzToken: AuthzToken
+               ): LiftResponse = {
+      (for {
+        techniqueId <- restExtractorService.extractString("techniqueId")(req)(Full(_)).toIO.notOptional("category parameter is missing")
```
```suggestion
        techniqueId <- restExtractorService.extractString("techniqueId")(req)(Full(_)).toIO.notOptional("technique id parameter is missing")
```

fanf commented on webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/ResourceFileService.scala:
```diff
@@ -115,4 +116,17 @@ class GitResourceFileService(gitReposProvider: GitRepositoryProvider) extends Re
+
+    IOResult.attempt(s"An error occurred while copying resources of technique ${techniqueId}/${techniqueVersion} to a new draft technique ${draftId}") {
+      val resourceDir = gitReposProvider.rootDirectory / s"techniques/${category}/${techniqueId}/${techniqueVersion}/resources"
+      val workspace = gitReposProvider.rootDirectory / s"workspace/${draftId}/${techniqueVersion}"
```
please avoid strings for global vars. Ideally they should be services for easier testing, or at lest something provided centrally in an object.

fanf commented on webapp/sources/rudder/rudder-web/src/main/elm/sources/Editor/DataTypes.elm:
```diff
@@ -240,6 +241,7 @@ type Msg =
   | UpdateTechnique Technique
   | DeleteTechnique (Result (Http.Detailed.Error String) ( Http.Metadata, TechniqueId ))
   | GetTechniqueResources  (Result (Http.Detailed.Error String) ( Http.Metadata, List Resource ))
+  | CopyResources  (Result (Http.Detailed.Error Bytes) ())
```
I don't get why you have bytes here ? The resources remains in the backend in the case of cloning no ?

VinceMacBuche replied:
because i don't care about the answer and that's how you do in Elm

fanf commented:
Can you explain the general idea here ? There's several things I don't get: 
- why does it need it's dedicated public API endpoint for something that looks like a very internal process implementation ? I would understand a general API "clone", but "copy resource when cloning" is really looking like an internal step, not an API usable by users
- why do you need Bytes in the elm part?
- I don't understand how cloning resources from a draft work either

VinceMacBuche commented:
Commit modified

VinceMacBuche commented:
Commit modified

fanf commented:
When I clone a technique with an existing resource, the resource is cloned. But as soons as I try to modify the technique name, I get a lot of error notif in UI, and the following stack trace in logs: 
```2024-05-03 12:54:28+0200 ERROR api-processing - SystemError: An error occurred while copying resources of technique upgrade_rudder_agent_on_debian_12_clone/1.0 to a new draft technique 4dda3477-60a0-4982-90b1-39c2e3f7a80f; cause was: java.nio.file.FileAlreadyExistsException: /var/rudder/configuration-repository-8.0/workspace/4dda3477-60a0-4982-90b1-39c2e3f7a80f/1.0/resources/rudder.list.bookworm 
 -> com.normation.rudder.ncf.GitResourceFileService.$anonfun$cloneResourcesFromTechnique$1(ResourceFileService.scala:147)
 -> com.normation.zio$ZioRuntime$.$anonfun$unsafeRun$1(ZioCommons.scala:445)
 -> com.normation.zio$ZioRuntime$.unsafeRun(ZioCommons.scala:445)
 -> com.normation.zio$ZioRuntime$.runNow(ZioCommons.scala:428)
 -> com.normation.zio$UnsafeRun.runNow(ZioCommons.scala:454)
 -> com.normation.rudder.rest.RudderJsonResponse$implicits$ToLiftResponseOne.toLiftResponseOne(RudderJsonResponse.scala:220)
 -> com.normation.rudder.rest.RudderJsonResponse$implicits$ToLiftResponseOne.toLiftResponseOne(RudderJsonResponse.scala:232)
 -> com.normation.rudder.rest.lift.TechniqueApi$CopyResourcesWhenCloning$.process(TechniqueApi.scala:235)
 -> com.normation.rudder.rest.lift.TechniqueApi$CopyResourcesWhenCloning$.process(TechniqueApi.scala:216)
 -> com.normation.rudder.rest.lift.LiftApiModule.handler(LiftApiDispatcher.scala:67)
 -> com.normation.rudder.rest.lift.LiftApiModule.handler$(LiftApiDispatcher.scala:59)
 -> com.normation.rudder.rest.lift.TechniqueApi$CopyResourcesWhenCloning$.handler(TechniqueApi.scala:216)
 -> com.normation.rudder.rest.lift.TechniqueApi$CopyResourcesWhenCloning$.handler(TechniqueApi.scala:216)
 -> com.normation.rudder.rest.BuildHandler.$anonfun$buildApi$21(ApiDatastructures.scala:696)
 -> com.normation.rudder.rest.BuildHandler.$anonfun$buildApi$18(ApiDatastructures.scala:680)
 -> com.normation.rudder.rest.BuildHandler.$anonfun$buildApi$15(ApiDatastructures.scala:674)
 -> com.normation.rudder.rest.BuildHandler.$anonfun$buildApi$15$adapted(ApiDatastructures.scala:672)```


Plus, the copy API it still public, it should be private. 

The clone of template with resources works great :+1:

VinceMacBuche commented:
Commit modified

VinceMacBuche commented:
I fixed the elm section, we were cloning resource at every edition, when we updated clone: https://github.com/Normation/rudder/pull/5453/files#diff-a4856d1cc4cf1a4d42d660a6e4c3f98072e5f685683419c57d1cd6f2612a0843R617 , it should only be done at clone ( https://github.com/Normation/rudder/pull/5453/files#diff-093c6a0e361570c9b6c31d97a4be6ef77d965334ffac9024abf42cf9badf9671R312 )

I set the api to private too https://github.com/Normation/rudder/pull/5453/files#diff-a4856d1cc4cf1a4d42d660a6e4c3f98072e5f685683419c57d1cd6f2612a0843R617

fanf reviewed (approved):
Great, works as expected, code is neat!

fanf commented:
OK, merging this PR
